{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from cyclomics import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##USER INPUTS\n",
    "\n",
    "#specify data folder\n",
    "#data from https://zenodo.org/record/3925250/files/Cyclomics_manuscript.zip\n",
    "data_folder = f'/Volumes/1TB/Cyclomics_manuscript/RCA'\n",
    "\n",
    "#samples to be processed\n",
    "samples = [\n",
    "    'CY_PJET_12WT_0001_000',\n",
    "    'CY_SS_PC_HC_0001_001_000',\n",
    "    'CY_SM_PC_HC_0002_001_000',\n",
    "    'CY_SM_PC_HC_0004_001_000',\n",
    "]\n",
    "\n",
    "min_cov = 10_000 #minimum number of consensus reads per position\n",
    "x = [0, 0.001, 0.005, 0.01, 0.05, 0.1, 0.5, 1, 10, 100] #FP% scale\n",
    "\n",
    "save_as = 'COSMIC_FP_multiple_samples.pdf' #final output (PNG or PDF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##Import the cosmic database\n",
    "#mind! The coord are 1-indexed\n",
    "cosmic_database = './db/TP53_COSMIC_mutations_Jun2018_with_genomic_coordinates.csv'\n",
    "db = pd.read_csv(cosmic_database).sort_values(by='GRCh37 coord. (chr:base)')\n",
    "db = db[['GRCh37 coord. (chr:base)', 'c.Mutation', 'Type']]\n",
    "\n",
    "#convert to POS 0-indexed\n",
    "db['REF'] = db.apply(get_ref, axis=1)\n",
    "db['POS'] = db.apply(get_pos, axis=1)-1\n",
    "\n",
    "#Get alleles on POS\n",
    "db['g.Ref_Allele'] = db.apply(get_wt_alleles, axis=1)\n",
    "db['g.Mutant_Allele'] = db.apply(get_mut_alleles, axis=1)\n",
    "\n",
    "db.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##Multiple sample COSMIC FP% plot\n",
    "\n",
    "def add_cosmic_mutations(row):\n",
    "    global db\n",
    "    cy_ref = row['REF']\n",
    "    cy_pos = row['POS']\n",
    "    cosmic_ref = set(db[db.POS == cy_pos]['g.Ref_Allele'])\n",
    "    \n",
    "    if cy_ref in cosmic_ref:\n",
    "        cosmic_mut = set(db[db.POS == cy_pos]['g.Mutant_Allele'])\n",
    "        return cosmic_mut\n",
    "    return set([np.nan])\n",
    "\n",
    "\n",
    "#Figure layout\n",
    "plt.style.use(style)\n",
    "plt.figure(figsize=(16,9))\n",
    "plt.rcParams.update({'font.size': 20, 'text.color': 'black'})\n",
    "\n",
    "\n",
    "#Process all samples\n",
    "ref_seq = pd.read_csv('./db/COSMIC_ref_bases.csv', sep='\\t')\n",
    "for i, sample in enumerate(samples):\n",
    "    for file in list_of_files(f'{data_folder}/{sample}', 'txt', recursive=1):\n",
    "        #Find sambamba_output_cosmic files\n",
    "        if 'consensus_sambamba_output_cosmic' in file and 'SPLIT' not in file:\n",
    "            df = pd.read_csv(file, sep='\\t')\n",
    "            print('processing', sample, end='...')\n",
    "            df['REF'] = ref_seq\n",
    "            df = df[df['COV'] >= min_cov]\n",
    "            df['COSMIC'] = df.apply(add_cosmic_mutations, axis=1)\n",
    "            df = df.drop(df[(df.COSMIC == {np.nan})].index) #Filter Nan\n",
    "            df['FP%_COSMIC'] = df.apply(fp_cosmic, axis=1)\n",
    "            y = []\n",
    "            for i, f in enumerate(x):\n",
    "                y.append(\n",
    "                    percent_of(\n",
    "                        len(df),\n",
    "                        len(df[df['FP%_COSMIC'] <= f])\n",
    "                    )\n",
    "                )\n",
    "\n",
    "            plt.plot(x,y, linewidth=4,label=sample)\n",
    "            print('OK')\n",
    "\n",
    "\n",
    "#Final plot\n",
    "plt.xscale('log')\n",
    "plt.ylim(min(x),max(x))\n",
    "plt.ylim(0,100)\n",
    "plt.xlabel('FP%')\n",
    "plt.ylabel('% bases')\n",
    "plt.title('COSMIC FP rate\\n')\n",
    "plt.gca().invert_xaxis()\n",
    "\n",
    "\n",
    "#Set proper labels\n",
    "plt.legend(loc='upper right',\n",
    "           labels=[\n",
    "               'single exon NO-PCR',\n",
    "               'single exon PCR from cfDNA',\n",
    "               'multiple exons PCR from cfDNA',\n",
    "               'all exons, panel from cfDNA'\n",
    "            ])\n",
    "\n",
    "\n",
    "#Save to PDF file\n",
    "plt.savefig(save_as, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##Same analysis with FOR/REV split\n",
    "save_as = 'COSMIC_FP_FR_multiple_samples.pdf' #final output (PNG or PDF)\n",
    "\n",
    "\n",
    "\n",
    "#Figure layout\n",
    "plt.style.use(style)\n",
    "plt.figure(figsize=(16,9))\n",
    "plt.rcParams.update({'font.size': 20, 'text.color': 'black'})\n",
    "\n",
    "#FOR-REV database\n",
    "fr_db = pd.read_csv('db/for_rev_positions.txt', sep='\\t')\n",
    "fr_db = fr_db[fr_db.REF == '17'][['REF','POS','type']]\n",
    "#print(len(fr_db))\n",
    "#fr_db.head()\n",
    "\n",
    "for i, sample in enumerate(samples):\n",
    "    \n",
    "    #Forward SPLIT\n",
    "    f_df = pd.read_csv(\n",
    "        f'{data_folder}/{sample}/for_rev_split/for_rev_split_forward_sambamba_output_cosmic.txt',\n",
    "        sep='\\t'\n",
    "    )\n",
    "    \n",
    "    #Reverse SPLIT\n",
    "    r_df = pd.read_csv(\n",
    "        f'{data_folder}/{sample}/for_rev_split/for_rev_split_reverse_sambamba_output_cosmic.txt',\n",
    "        sep='\\t'\n",
    "    )\n",
    "    \n",
    "    for file in list_of_files(f'{data_folder}/{sample}', 'txt', recursive=1):\n",
    "        #Find sambamba_output_cosmic files\n",
    "        if 'consensus_sambamba_output_cosmic' in file and 'SPLIT' not in file:\n",
    "            df = pd.read_csv(file, sep='\\t')\n",
    "            print('processing', sample, end='...')\n",
    "            \n",
    "            #Patch df with FOR/REV data\n",
    "            for i,row in fr_db.iterrows():\n",
    "                if row.type == 'FOR':\n",
    "                    data = f_df[(f_df.REF == int(row.REF)) & (f_df.POS == int(row.POS))]\n",
    "                    \n",
    "\n",
    "                elif row.type == 'REV':\n",
    "                    data = r_df[(r_df.REF == int(row.REF)) & (r_df.POS == int(row.POS))]\n",
    "                df = df.append(data)\n",
    "                \n",
    "            df = df.drop_duplicates(subset=['REF','POS'], keep='last')\n",
    "            df = df.sort_values(['POS'])\n",
    "            df = df.reset_index(drop=True)\n",
    "            \n",
    "            df['REF'] = ref_seq\n",
    "            df = df[df['COV'] >= min_cov]\n",
    "            df['COSMIC'] = df.apply(add_cosmic_mutations, axis=1)\n",
    "            df = df.drop(df[(df.COSMIC == {np.nan})].index) #Filter Nan\n",
    "            df['FP%_COSMIC'] = df.apply(fp_cosmic, axis=1)\n",
    "            \n",
    "            y = []\n",
    "            for i, f in enumerate(x):\n",
    "                y.append(\n",
    "                    percent_of(\n",
    "                        len(df),\n",
    "                        len(df[df['FP%_COSMIC'] <= f])\n",
    "                    )\n",
    "                )\n",
    "\n",
    "            plt.plot(x,y, linewidth=4,label=sample)\n",
    "            print('OK')\n",
    "\n",
    "\n",
    "#Final plot\n",
    "plt.xscale('log')\n",
    "plt.ylim(min(x),max(x))\n",
    "plt.ylim(0,100)\n",
    "plt.xlabel('FP%')\n",
    "plt.ylabel('% bases')\n",
    "plt.title('COSMIC FP rate\\n')\n",
    "plt.gca().invert_xaxis()\n",
    "\n",
    "\n",
    "#Set proper labels\n",
    "plt.legend(loc='upper right',\n",
    "           labels=[\n",
    "               'single exon NO-PCR',\n",
    "               'single exon PCR from cfDNA',\n",
    "               'multiple exons PCR from cfDNA',\n",
    "               'all exons, panel from cfDNA'\n",
    "            ])\n",
    "\n",
    "\n",
    "#Save to PDF file\n",
    "plt.savefig(save_as, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
