{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-11-19T08:47:22.042132Z",
     "start_time": "2019-11-19T08:47:20.038221Z"
    }
   },
   "outputs": [],
   "source": [
    "from cyclomics import *\n",
    "warnings.simplefilter(\"ignore\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##USER INPUTS\n",
    "\n",
    "#specify data folder\n",
    "#data from https://zenodo.org/record/3925250/files/Cyclomics_manuscript.zip\n",
    "data_folder = f'/Volumes/1TB/Cyclomics_manuscript/RCA'\n",
    "\n",
    "#sample to process\n",
    "sample_name = 'CY_LOT1_QC_0001_000'\n",
    "\n",
    "\n",
    "show_plots   = True #show plots in this notebook (it can consume lots of RAM)\n",
    "\n",
    "#Save plots to drive\n",
    "save_fig     = True\n",
    "save_in      = 'plots'\n",
    "\n",
    "\n",
    "#Basic filtering\n",
    "read_all = True #if True, read all the file (mind not running out of RAM)\n",
    "lines_to_read = 50_000 #if not read_all, choose how many lines to read (1 read per line)\n",
    "len_filter = 5_000 #bases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-11-19T08:50:15.081754Z",
     "start_time": "2019-11-19T08:50:12.738421Z"
    }
   },
   "outputs": [],
   "source": [
    "##Run\n",
    "sample_folder = f'{data_folder}/{sample_name}/'\n",
    "\n",
    "if save_fig:\n",
    "    os.mkdir(f'{sample_folder}/{save_in}')\n",
    "    #!mkdir $sample_folder/$save_in\n",
    "\n",
    "files = list_of_files(sample_folder, 'txt.gz', recursive=1)\n",
    "structure_file = [f for f in files if 'structure.txt.gz' in f][0]\n",
    "\n",
    "\n",
    "\n",
    "data = []\n",
    "read_len_dist     = Counter() #all reads here\n",
    "composition_dict  = {}        #all reads passing len_filter here\n",
    "\n",
    "with gzip.open(structure_file, 'r') as f:\n",
    "    line_number = 0\n",
    "    for line_number, line in enumerate(f):\n",
    "        if line_number == 0:\n",
    "            continue\n",
    "        #if not line.startswith('/'):\n",
    "        #    continue\n",
    "            \n",
    "        d = line.decode()strip().split('\\t')\n",
    "        try:\n",
    "            if d[1] == 'nan':\n",
    "                read_len_dist.update({0})\n",
    "                continue\n",
    "        except Exception as e:\n",
    "            print(e)\n",
    "            print(line)\n",
    "            print(line_number)\n",
    "            break\n",
    "            \n",
    "        if len(d) > 1: #and line.startswith('/'):\n",
    "            \n",
    "            data.append(d[:2])\n",
    "            read_len = sum([int(i.split(':')[0]) for i in d[1].split(',')])\n",
    "            read_len_dist.update({read_len})\n",
    "            \n",
    "            if read_len >= len_filter:\n",
    "                segments        = d[1].split(',')\n",
    "                composition     = Counter([i.split(':')[1] for i in segments])\n",
    "                \n",
    "                for k in ['I','BB','U']:\n",
    "                    if k not in composition:\n",
    "                        composition[k] = 0\n",
    "                composition_dict[d[0]] = composition\n",
    "    \n",
    "        else:\n",
    "            read_len_dist.update({0})\n",
    "        \n",
    "        if not read_all:\n",
    "            lines_to_read -= 1\n",
    "            if not lines_to_read:\n",
    "                break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-11-19T08:50:27.738769Z",
     "start_time": "2019-11-19T08:50:25.149195Z"
    }
   },
   "outputs": [],
   "source": [
    "##Colormap normalization based on TP53 coordinates 17:7565097-7590856 (GRCh37)\n",
    "\n",
    "#https://matplotlib.org/users/colormaps.html\n",
    "cmap_0 = cm.gist_rainbow\n",
    "cmap_1 = cm.prism #sharper difference between close numbers but not unique\n",
    "cmap_2 = cm.tab20\n",
    "cmap_chr = cm.hsv #for chromosomes\n",
    "\n",
    "##From Cyclomics_TP53_assay.bed\n",
    "#17\t7572800\t7573100\n",
    "#17\t7573800\t7574100\n",
    "#17\t7576400\t7577700\n",
    "#17\t7578100\t7578600\n",
    "#17\t7579200\t7580000\n",
    "\n",
    "vmin=7572800 #TP53 mapping start\n",
    "vmax=7580000 #TP53 mapping end\n",
    "\n",
    "#normalize coordinates\n",
    "norm = Normalize(vmin, vmax)\n",
    "norm_chr = Normalize(1, 24) #for chromosomes\n",
    "\n",
    "\n",
    "##Chr color map\n",
    "plt.style.use('ggplot')\n",
    "plt.figure(figsize=(20,4), dpi=200)\n",
    "y = 0\n",
    "r = list(range(1, 25))\n",
    "for x in r:\n",
    "    plt.scatter(x,y  , s=300, color=cmap_chr(norm_chr(x)), lw=1, edgecolors='black')\n",
    "plt.xticks(r)\n",
    "plt.title('Chromosomes color map')\n",
    "plt.show()\n",
    "\n",
    "\n",
    "##TP53 color cmap\n",
    "#for style in plt.style.available:\n",
    "plt.style.use('ggplot')\n",
    "plt.figure(figsize=(20,4), dpi=200)\n",
    "y = 0\n",
    "r = range(vmin, vmax, 100)\n",
    "for x in r:\n",
    "    plt.scatter(x,y+1, s=100, color=cmap_0(norm(x)), lw=1, edgecolors='black')\n",
    "    plt.scatter(x,y  , s=100, color=cmap_1(norm(x)), lw=1, edgecolors='black')\n",
    "    plt.scatter(x,y-1, s=100, color=cmap_2(norm(x)), lw=1, edgecolors='black')\n",
    "plt.title('TP53 color map')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-11-19T08:56:44.975588Z",
     "start_time": "2019-11-19T08:55:24.211107Z"
    },
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "## Example\n",
    "#structure = '63:U,188:BB:R:BB200_2:1:194:0,20:U,21:U,...,195:BB:R:BB200_2:2:194:0,182:U'\n",
    "\n",
    "\n",
    "insert = 'I'\n",
    "#backbones color chart\n",
    "bb_colors = {\n",
    "    'BB22':'RED',\n",
    "    'BB24':'BLUE',\n",
    "    'BB25':'GRAY',\n",
    "    'BBCR':'ORANGE',\n",
    "    'PJET':'GREEN'\n",
    "}\n",
    "\n",
    "with_multiple_BB = []\n",
    "with_multiple_I = []\n",
    "with_template_shift = {\"None\":[],\n",
    "                       \"Unexpected\":[],\n",
    "                       \"Regular\":[],\n",
    "                       \"??\":[]}\n",
    "\n",
    "\n",
    "#TP53 on GRCh37\n",
    "#expected_chr = '17'\n",
    "#vmin=7572800 #TP53 mapping start\n",
    "#vmax=7580000 #TP53 mapping end\n",
    "\n",
    "\n",
    "\n",
    "#TP53 sequence only\n",
    "expected_chr = 'TP53'\n",
    "vmin=0 #TP53 mapping start\n",
    "vmax=25772 #TP53 mapping end\n",
    "\n",
    "\n",
    "#BRAF on GRCh37\n",
    "#expected_chr = '7'\n",
    "#vmin = 140452000\n",
    "#vmax = 140458000\n",
    "\n",
    "add_label = False\n",
    "label = None #defaul label\n",
    "\n",
    "try:\n",
    "    if not len(processed): #skip if already processed, added to resume the labeling\n",
    "        label_dict = {0:'discard',\n",
    "                      1:'keep_all',\n",
    "                      2:'keep_longest_block',\n",
    "                      3:'keep_most_common_insert',\n",
    "                      4:'complex_fix_needed',\n",
    "                      5:'remap',\n",
    "                      6:'??'}\n",
    "        label_string = '|'.join([f'{k}:{v}'for k,v in label_dict.items()])\n",
    "        grouped_by_labels = {}\n",
    "except NameError:\n",
    "    processed = set()\n",
    "\n",
    "actions = {} #for filtering debug    \n",
    "    \n",
    "#Filter reads\n",
    "print_report = True\n",
    "filter1_on    = True #Optimal reads\n",
    "filter2_on    = False\n",
    "filter3_on    = False\n",
    "filter4_on    = False #BB only\n",
    "\n",
    "min_repeats_filter = 5 #number of BB-I blocks\n",
    "chr_filter_in = 'TP53'\n",
    "chr_filter_out = 'BRAF'\n",
    "max_unmapped = 500\n",
    "filter_pass = [] \n",
    "\n",
    "\n",
    "insertID_count = Counter()\n",
    "\n",
    "\n",
    "for n, (bam, structure) in enumerate(data):#[:1000]):\n",
    "    \n",
    "    bam = bam.split('/')[-1]\n",
    "    if bam in processed:\n",
    "        continue\n",
    "\n",
    "    ##Desecting\n",
    "    segments        = structure.split(',')\n",
    "    composition     = Counter([i.split(':')[1] for i in segments])\n",
    "    seg_len = segments_len(structure)\n",
    " \n",
    "    mapped      = Counter(i.split(':')[3] for i in segments if i[-2:] !=':U')\n",
    "    unmapped    = Counter('U' for i in segments if i[-2:] ==':U')\n",
    "    forward_BB  = structure.count('BB:F')\n",
    "    reverse_BB  = structure.count('BB:R')\n",
    "    longer_unmapped_region = max(int(i.split(':')[0]) for i in segments if i[-2:] ==':U')    \n",
    "    c_str           = collapse(structure, remove_dust=25, trim_edges=1) #collapsed structure\n",
    "    c_blocks        = list(get_reps(c_str[0]))  #consecutive blocks\n",
    "\n",
    "    dna_shift       = check_dna_shift(c_str, c_blocks)\n",
    "    with_template_shift[dna_shift].append(bam)\n",
    "\n",
    "    Iid             = Counter()\n",
    "    BBid            = Counter()\n",
    "    read_len        = 0\n",
    "    for s in segments:\n",
    "        i = s.split(':')\n",
    "        if i[1] == 'I':\n",
    "            Iid.update({i[-1]})\n",
    "            \n",
    "        elif i[1] == 'BB':\n",
    "            BBid.update({i[-1]})\n",
    "        read_len += int(i[0])\n",
    "\n",
    "    if len(Iid) > 1:\n",
    "        sorted_v = sorted(Iid.values()) \n",
    "        if sorted_v[0] > 5 and sorted_v[0]/sorted_v[0] < 3:\n",
    "            insertID_count.update({len(Iid)})\n",
    "        else:\n",
    "            insertID_count.update({1})\n",
    "    else:\n",
    "        insertID_count.update({1})\n",
    "\n",
    "\n",
    "    if len(BBid) > 1:\n",
    "        with_multiple_BB.append(bam)\n",
    "    if len(Iid) > 1:\n",
    "        with_multiple_I.append(bam)\n",
    "\n",
    "\n",
    "    ##Scoring\n",
    "    len_s = len(segments)\n",
    "\n",
    "\n",
    "    ##Filtering#########\n",
    "    \n",
    "    ##Filter 1##\n",
    "    if filter1_on:\n",
    "        if longer_unmapped_region > max_unmapped:\n",
    "            continue\n",
    "        if not mapped:\n",
    "            continue\n",
    "        elif chr_filter_in not in mapped or chr_filter_out in mapped:\n",
    "            continue\n",
    "        if 'I' not in composition or 'BB' not in composition:\n",
    "            continue\n",
    "        elif composition['I'] < min_repeats_filter or composition['BB'] < min_repeats_filter:\n",
    "            continue\n",
    "        elif not 0.75 <= (composition['I']/composition['BB']) <= 1.25:\n",
    "            continue\n",
    "        else:\n",
    "            filter_pass.append(bam)\n",
    "    ####################\n",
    "    \n",
    "    \n",
    "    ##Filter 2##\n",
    "    if filter2_on:\n",
    "        if longer_unmapped_region > max_unmapped:\n",
    "            continue\n",
    "        elif composition['BB'] < min_repeats_filter:\n",
    "            continue\n",
    "        #only forward or only reverse BB in one read\n",
    "        if forward_BB and reverse_BB:\n",
    "            continue\n",
    "        else:\n",
    "            filter_pass.append(bam)\n",
    "    ####################\n",
    "    \n",
    "    \n",
    "    ##Filter 3##\n",
    "    if filter3_on:\n",
    "        if 'I' not in composition or 'BB' not in composition:\n",
    "            continue\n",
    "        if seg_len['I'] < 100: #Inserts of at least 100bp\n",
    "            continue\n",
    "        if seg_len['BB'] < 200: #BB of at least 200bp\n",
    "            continue\n",
    "        else:\n",
    "            filter_pass.append(bam)\n",
    "    ####################\n",
    "    \n",
    "    \n",
    "    ##Filter 4##\n",
    "    if filter4_on:\n",
    "        if 'BB' not in composition or 'I' in composition:\n",
    "            continue\n",
    "        if seg_len['BB'] < 200: #BB of at least 200bp\n",
    "            continue\n",
    "        else:\n",
    "            filter_pass.append(bam)\n",
    "    ####################\n",
    "    \n",
    "\n",
    "\n",
    "    if show_plots:\n",
    "        plot_structure(structure,\n",
    "                       bb_colors,\n",
    "                       insert,\n",
    "                       expected_chr,\n",
    "                       vmin,\n",
    "                       vmax,\n",
    "                       title=bam,\n",
    "                       show=show_plots,\n",
    "                       savefig=f'{sample_folder}{save_in}/{bam}.png' if save_fig else False\n",
    "                      )\n",
    "\n",
    "    if add_label:\n",
    "        while True:\n",
    "            label = int(input(f'Input one of the following labels:\\n{label_string}\\n'))\n",
    "            if label in label_dict:\n",
    "                break\n",
    "            else:\n",
    "                print('Wrong label value!')\n",
    "        try:\n",
    "            grouped_by_labels[label].append(bam)\n",
    "        except KeyError:\n",
    "            grouped_by_labels[label] = [bam]\n",
    "\n",
    "\n",
    "\n",
    "    ##Filtering demo\n",
    "    action = 'discard'\n",
    "    if 'I' in composition and composition['I'] >= 5:\n",
    "        action = 'remap'\n",
    "        if 'BB' in composition:\n",
    "            if 0.8 <= (composition['I']/composition['BB']) <= 1.4:\n",
    "                action = 'keep'\n",
    "\n",
    "    try:\n",
    "        actions[action].append(bam)\n",
    "    except KeyError:\n",
    "        actions[action] = [bam]\n",
    "\n",
    "    ##Report\n",
    "    if print_report:\n",
    "        print(f'seqence N    : {n}')\n",
    "        print(f'label        : {label}')\n",
    "        print(f'from         : {bam}')\n",
    "        print(f'read length  : {read_len}')\n",
    "        ####\n",
    "        #Add alignment of consensus vs target mutation\n",
    "        #Flag mutants expected/unexpected contaminant\n",
    "\n",
    "        ####\n",
    "        print(f'structure    : {structure}')\n",
    "        print(f'segments_len : {seg_len}')\n",
    "        print(f'c_structure  : {c_str}')\n",
    "        print(f'c_blocks     : {c_blocks}')\n",
    "        print(f'dna_shift    : {dna_shift}')\n",
    "        print(f'segments     : {len(segments)}')\n",
    "        print(f'composition  : {composition}')\n",
    "        print(f'Insert   IDs : {Iid}')\n",
    "        print(f'Backbone IDs : {BBid}')\n",
    "        print(f'mapped       : {mapped}')\n",
    "        print(f'unmapped     : {unmapped}')\n",
    "        print(f'action       : {action}')\n",
    "        print(f'\\n{\"-\"*20}\\n')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
